datashape/discovery.py

```diff
@@ -396,3 +396,29 @@ def descendents(d, x):
         children -= desc
         desc.update(children)
     return desc
+
+
+def _tp_eq(a, b):
+    return a == b or b is null
```

Comment:
I don't think we want to raise on NULL, the idea is that slice(None, 5) should be a range of `int` even though one side is unbound.